Block publisher.playground/Sources/BlockPublisher.swift

```diff
@@ -0,0 +1,68 @@
+import Combine
+
+public typealias CancelClosure = () -> Void
+
+public extension Publishers {
+    struct BlockPublisher<Output, Failure: Error>: Publisher {
+        private let creationClosure: (BlockSubscription) -> CancelClosure
+        
+        public init(_ creationClosure: @escaping (BlockSubscription) -> CancelClosure) {
+            self.creationClosure = creationClosure
+        }
+        
+        public func receive<S>(subscriber: S) where S : Subscriber, Failure == S.Failure, Output == S.Input {
+            let subscription = BlockSubscription(subscriber: AnySubscriber(subscriber))
+            subscriber.receive(subscription: subscription)
+            let action = creationClosure(subscription)
```

Comment:
The assumption is the following (verified with the debugger): 
when you deliver a subscription to the subscriber, he responds synchronously with the demand.
Then before line line 16 I have the demand stored in the subscription.
Synchronous cancel doesn't seem to work even with Combine publishers and Sink subscriber:

Run the following:
```
[1,2,3,5].publisher
            .delay(for: 2, scheduler: RunLoop.main)
            .sink(receiveCompletion: { print($0) }, receiveValue: { print($0) })
            .cancel()
```

Reply:
I'm scared about the word "assumption". It's like that right now but it might be different in future or also there might be a custom Subscriber implementation that does not follow this rule.

Reply:
It could be...
But to be honest I'm not able to think a reason why a subscriber should delay the `request(_: Demand)`message.
Anyway the order of the messages is the following: 

<img width="1121" alt="Schermata 2019-10-06 alle 14 00 21" src="https://user-images.githubusercontent.com/19324622/66268844-c066ed80-e841-11e9-89b8-e2f4d1b4a85d.png">

Reply:
No idea why it should, I'm pointing to the the fact that might be.
The diagram is super clear here, I agree, but it describe who ask for what but it does not describe the implementation detail. 
That kind of diagrams in general best describe asynchronous flows, the more down you go the more time is passed. So `request(_: Demand)` might also be interpreted as asynchronous in general.
As we saw in some implementations the Subscription takes a key role and it's barely mentioned here.

Again, being talking about an High Level Framework to accomplish asynchronoous task, I would never assume that something happen synchronously.

Reply:
I understand your point.
I think the fix is actually simple: just defer the execution of the line
`creationClosure(subscription)` inside `request(_ demand:)`!  ;-)

Reply:
Checkout the commit:
b35dcf23a73c0a417b6ca3090a77f0c7aed6a86a